starfish/core/util/dtype.py

```diff
@@ -3,40 +3,95 @@
 import numpy as np
 import xarray as xr
 
+
 def preserve_float_range(
         array: Union[xr.DataArray, np.ndarray],
-        rescale: bool=False) -> Union[xr.DataArray, np.ndarray]:
+        rescale: bool = False,
+        preserve_input: bool = False,
+) -> Union[xr.DataArray, np.ndarray]:
     """
     Clips values below zero to zero. If values above one are detected, clips them
     to 1 unless `rescale` is True, in which case the input is scaled by
     the max value and the dynamic range is preserved.
 
+    Input arrays may be modified by this operation if `preserve_input` is False.
+
     Parameters
     ----------
     array : Union[xr.DataArray, np.ndarray]
         Array whose values should be in the interval [0, 1] but may not be.
-    rescale: bool
+    rescale : bool
         If true, scale values by the max.
+    preserve_input : bool
+        if True, ensure that we do not modify the input data by making a copy.
 
     Returns
     -------
     array : Union[xr.DataArray, np.ndarray]
         Array whose values are in the interval [0, 1].
 
     """
-    array = array.copy()
-
     if isinstance(array, xr.DataArray):
+        # do a shallow copy
+        array = array.copy(deep=False)
         data = array.values
     else:
         data = array
 
-    negative = data < 0
-    if np.any(negative):
-        data[negative] = 0
-    if np.any(data > 1):
-        if rescale:
-            data /= data.max()
-        else:
-            data[data > 1] = 1
-    return array.astype(np.float32)
+    casted = data.astype(np.float32, copy=False)
+    if casted is not data:
+        preserve_input = False
+    data = casted
+
+    if preserve_input or not data.flags['WRITEABLE']:
```

Comment:
`if not data.flags['WRITEABLE']` I think this should emit a warning, as the function will do something different than the user expects.

Reply:
How is that?

Reply:
What would happen if they said `preserve_input=False`, but `data.flags['WRITEABLE']` was also False?

Reply:
We'd make a copy.  `preserve_input==False` does not mean we're not allowed to make a copy.

Reply:
Got it. I understand.